File: examples/csharp/Table/SagaHand/TableHandSaga.cs
// DOC: This file is referenced in docs/docs/examples/sagas.mdx
//      Update documentation when making changes to saga patterns.

using Angzarr;
using Angzarr.Client;
using Angzarr.Examples;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;

namespace Table.SagaHand;

/// <summary>
/// Saga: Table -> Hand
/// Reacts to HandStarted events from Table domain.
/// Sends DealCards commands to Hand domain.
/// Sagas are stateless translators - framework handles sequence stamping.
/// </summary>
public static class TableHandSaga
{
    // docs:start:event_router
    public static EventRouter Create()
    {
        return new EventRouter("saga-table-hand")
            .Domain("table")
            .On<HandStarted>(HandleHandStarted);
    }

    // docs:end:event_router

    // docs:start:saga_handler
    private static object HandleHandStarted(HandStarted evt, List<EventBook> destinations)
    {
        // Sagas are stateless - destinations not used, framework stamps sequences
        var players = evt
            .ActivePlayers.Select(seat => new PlayerInHand
            {
                PlayerRoot = seat.PlayerRoot,
                Position = seat.Position,
                Stack = seat.Stack,
            })
            .ToList();

        var dealCards = new DealCards
        {
            TableRoot = evt.HandRoot,
            HandNumber = evt.HandNumber,
            GameVariant = evt.GameVariant,
            DealerPosition = evt.DealerPosition,
            SmallBlind = evt.SmallBlind,
            BigBlind = evt.BigBlind,
        };
        dealCards.Players.AddRange(players);

        var cmdAny = EventRouter.PackCommand(dealCards);

        return new CommandBook
        {
            Cover = new Cover
            {
                Domain = "hand",
                Root = new UUID { Value = evt.HandRoot },
            },
            Pages =
            {
                new CommandPage
                {
                    Header = new PageHeader { AngzarrDeferred = new AngzarrDeferredSequence() },
                    Command = cmdAny,
                },
            },
        };
    }
    // docs:end:saga_handler
}

File: examples/cpp/table/saga-hand/src/table_hand_saga.cpp
// DOC: This file is referenced in docs/docs/examples/sagas.mdx
//      Update documentation when making changes to saga patterns.

#include "table_hand_saga.hpp"

namespace table {
namespace saga {

// docs:start:saga_handler
// Handle HandStarted: produce DealCards command for hand.
// Sagas are stateless translators - framework handles sequence stamping.
static std::vector<angzarr::CommandBook> handle_hand_started(
    const google::protobuf::Any& event_any, const std::string& source_root,
    const std::string& correlation_id, const std::vector<angzarr::EventBook>& destinations) {
    (void)source_root;
    (void)destinations;  // Sagas are stateless - destinations not used

    examples::HandStarted event;
    event_any.UnpackTo(&event);

    // Build DealCards command from HandStarted event
    examples::DealCards deal_cards;
    deal_cards.set_table_root(event.hand_root());
    deal_cards.set_hand_number(event.hand_number());
    deal_cards.set_game_variant(event.game_variant());
    deal_cards.set_dealer_position(event.dealer_position());
    deal_cards.set_small_blind(event.small_blind());
    deal_cards.set_big_blind(event.big_blind());

    // Add players from active players
    for (const auto& seat : event.active_players()) {
        auto* player = deal_cards.add_players();
        player->set_player_root(seat.player_root());
        player->set_position(seat.position());
        player->set_stack(seat.stack());
    }

    // Pack command
    google::protobuf::Any cmd_any;
    cmd_any.PackFrom(deal_cards, "type.googleapis.com/");

    // Build command book
    angzarr::CommandBook cmd_book;
    cmd_book.mutable_cover()->set_domain("hand");
    cmd_book.mutable_cover()->mutable_root()->set_value(event.hand_root());
    cmd_book.mutable_cover()->set_correlation_id(correlation_id);

    auto* page = cmd_book.add_pages();
    // Framework handles sequence stamping
    page->mutable_header()->mutable_angzarr_deferred();
    page->mutable_command()->CopyFrom(cmd_any);

    return {std::move(cmd_book)};
}
// docs:end:saga_handler

// docs:start:event_router
angzarr::EventRouter create_table_hand_router() {
    return angzarr::EventRouter("saga-table-hand")
        .domain("table")
        .on("HandStarted", handle_hand_started);
}
// docs:end:event_router

}  // namespace saga
}  // namespace table
